src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs

```diff
@@ -0,0 +1,357 @@
+                    {
+                        Marshal.Copy(byteBuffer, 0, ptr, numBytes);
+                    }
+                    return (size_t)numBytes;
+                });
+        }
+
+        private static int CurlSendIoCtlCallback(
+            IntPtr handle,
+            int cmd,
+            IntPtr context)
+        {
+            if (cmd != curliocmd.CURLIOCMD_RESTARTREAD)
+            {
+                return curlioerr.CURLIOE_UNKNOWNCMD;
+            }
+
+            GCHandle gch = GCHandle.FromIntPtr(context);
+            var state = (RequestCompletionSource)gch.Target;
+
+            try
+            {
+                if (state.CancellationToken.IsCancellationRequested)
+                {
+                    HandleAsyncException(state, null);
+                    // Returing a value other than CURLIOE_OK fails the callback and forces
+                    // request completion with an error
+                    return curlioerr.CURLIOE_FAILRESTART;
+                }
+
+                Stream contentStream = state.RequestContentStream;
+                if (!contentStream.CanSeek)
+                {
+                    return curlioerr.CURLIOE_FAILRESTART;
+                }
+                contentStream.Seek(0, SeekOrigin.Begin);
+                return curlioerr.CURLIOE_OK;
+            }
+            catch (Exception ex)
+            {
+                HandleAsyncException(state, ex);
+                // Returing a value other than CURLIOE_OK fails the callback and forces
+                // request completion with an error
+                return curlioerr.CURLIOE_FAILRESTART;
+            }
+        }
+
+        private static size_t ExecuteCallback(
+            IntPtr buffer,
+            size_t size,
+            size_t nitems,
+            IntPtr context,
+            Func<IntPtr, size_t, RequestCompletionSource, size_t> callback)
+        {
+            size *= nitems;
+            GCHandle gch = GCHandle.FromIntPtr(context);
+            var state = (RequestCompletionSource)gch.Target;
+
+            try
+            {
+                return (size == 0) ? 0 : callback(buffer, size, state);
+            }
+            catch (Exception ex)
+            {
+                HandleAsyncException(state, ex);
+                // Returing a value other than size fails the callback and forces
+                // request completion with an error
+                return (size > 0) ? size - 1 : 1;
+            }
+        }
+
+        private static bool TryParseStatusLine(HttpResponseMessage response, string responseHeader)
+        {
+            if (!responseHeader.StartsWith(s_httpPrefix, StringComparison.OrdinalIgnoreCase))
+            {
+                return false;
+            }
+
+            // Check if line begins with HTTP/1.1 or HTTP/1.0
+            int prefixLength = s_httpPrefix.Length;
+            if ((responseHeader[prefixLength] == '1') && (responseHeader[prefixLength + 1] == '.'))
```

Comment:
What if responseHeader.Length == prefixLength such that prefixLength + 1 (or +2 as used below) is out of bounds?